File: android/app/src/main/java/org/celo/mobile/CapsuleSignerModule.java
package org.celo.mobile;

import android.util.Log;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import signer.Signer;

public class CapsuleSignerModule extends ReactContextBaseJavaModule {
  static final String TAG = "CapsuleSignerModule";

  String ids = "[\"USER\",\"RECOVERY\",\"CAPSULE\"]";
  int threshold = 2;
  String serverUrl = "http://mpcnetworkloadbalancer-348316826.us-west-1.elb.amazonaws.com";
  String configBase =
    "{\"ServerUrl\": \"%s\", \"WalletId\": \"%s\", \"Id\":\"%s\", \"Ids\":%s, \"Threshold\":1}";

  CapsuleSignerModule(ReactApplicationContext context) {
    super(context);
  }

  @Override
  public String getName() {
    return "CapsuleSignerModule";
  }

  /**
   * Perform distributed key generation with the Capsule server
   *
   * @param protocolId
   * @return
   */
  @ReactMethod
  public void createAccount(String walletId, String protocolId, String id, Promise promise) {
    String signerConfig = String.format(configBase, serverUrl, walletId, id, ids);
    (
      new Thread(
        () -> {
          String res = Signer.createAccount(serverUrl, signerConfig, protocolId);
          promise.resolve(res);
        }
      )
    ).start();
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public String getAddress(String serializedSigner) {
    return Signer.getAddress(serializedSigner);
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public String sendTransaction(String protocolId, String serializedSigner, String transaction) {
    return Signer.sendTransaction(serverUrl, serializedSigner, "hello world", protocolId);
  }
}

File: ios/CapsuleSignerModule.m
//
//  CapsuleSignerModule.m
//
//  Created by Michał Osadnik on 14/12/2022.
//

#import <Foundation/Foundation.h>
#import "CapsuleSignerModule.h"
#import <Signer/Signer.h>

static NSString *configBase = @"{\"ServerUrl\": \"%@\", \"WalletId\": \"%@\", \"Id\":\"%@\", \"Ids\":%@, \"Threshold\":1}";

static NSString *ids = @"[\"USER\",\"CAPSULE\",\"RECOVERY\"]";

static NSString *serverUrl = @"http://mpcnetworkloadbalancer-348316826.us-west-1.elb.amazonaws.com";


@implementation CapsuleSignerModule
+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

- (dispatch_queue_t)methodQueue {
  return dispatch_get_main_queue();
}

RCT_EXPORT_MODULE();

// Get Address
- (void) invokeSignerGetAddress:(NSDictionary*)params
{
  NSString* serializedSigner = [params objectForKey:@"serializedSigner"];
  RCTPromiseResolveBlock resolve = [params objectForKey:@"resolve"];

  NSString* res = SignerGetAddress(serializedSigner);
  resolve(res);
}

RCT_EXPORT_METHOD(getAddress:(NSString *)serializedSigner
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  NSDictionary* params = [NSDictionary dictionaryWithObjectsAndKeys:
                          resolve, @"resolve",
                          serializedSigner, @"serializedSigner",
                          nil];
  [self performSelectorInBackground:@selector(invokeSignerGetAddress:)
                         withObject:params];
}

// Send Transaction
- (void) invokeSignerSendTransaction:(NSDictionary*)params
{
  NSString* serializedSigner = [params objectForKey:@"serializedSigner"];
  NSString* transaction = [params objectForKey:@"transaction"];
  NSString* protocolId = [params objectForKey:@"protocolId"];
  RCTPromiseResolveBlock resolve = [params objectForKey:@"resolve"];
  
  NSString* res = SignerSendTransaction(serverUrl, serializedSigner, transaction, protocolId);
  resolve(res);
}

RCT_EXPORT_METHOD(sendTransaction:(NSString*)protocolId
                  serializedSigner:(NSString *)serializedSigner
                  transaction:(NSString *)transaction
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

{
  NSDictionary* params = [NSDictionary dictionaryWithObjectsAndKeys:
                          resolve, @"resolve",
                          serializedSigner, @"serializedSigner",
                          transaction, @"transaction",
                          protocolId, @"protocolId",
                          nil];
  [self performSelectorInBackground:@selector(invokeSignerSendTransaction:)
                         withObject:params];
}

// Create Account
- (void) invokeSignerCreateAccount:(NSDictionary*)params
{
  NSString* protocolId = [params objectForKey:@"protocolId"];
  NSString* signerConfig = [params objectForKey:@"signerConfig"];
  RCTPromiseResolveBlock resolve = [params objectForKey:@"resolve"];
  NSString* res = SignerCreateAccount(serverUrl,signerConfig, protocolId);
  
  resolve(res);
}

RCT_EXPORT_METHOD(createAccount:(NSString *)walletId
                  protocolId:(NSString *)protocolId
                  userId:(NSString *)userId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  NSString* signerConfig = [NSString stringWithFormat: configBase, serverUrl, walletId, userId, ids];
  NSDictionary* params = [NSDictionary dictionaryWithObjectsAndKeys:
                          protocolId, @"protocolId",
                          resolve, @"resolve",
                          signerConfig, @"signerConfig",
                          nil];
  [self performSelectorInBackground:@selector(invokeSignerCreateAccount:)
                         withObject:params];
}

@end
